### ISSUE TITLE
"spacy pretrain" vs huggingface pretrain

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#339]
## Context
Currently we use the scripts of [`data_and_models/pipelines/sentence_embedding/scripts/`](https://github.com/BlueBrain/Search/tree/master/data_and_models/pipelines/sentence_embedding/scripts) to pre-train the weights of the transformer model that is then used as a backbone and fine-tuned to generate our NER model. Those scripts are mainly  (a) a copy-paste of the `huggingface` script to run the LM (language modelling) pre-training ([here](https://github.com/huggingface/transformers/blob/v3.4.0/examples/language-modeling/run_language_modeling.py)) and (b) of  the `sentence-transformers` script to run the NLI+STS fine-tuning ([here](https://github.com/UKPLab/sentence-transformers/blob/v0.3.9/examples/training/other/training_multi-task.py)). 
But there's also a [`spacy pretrain`](https://spacy.io/api/cli#pretrain) that lets you pre-train a model using a self-superivised LM task ("LMAO"–Language Modelling with Approximate Outputs). The docs say that this allows to pre-train a `Tok2Vec` component, but it's not clear if they also support doing it for the alternative `Transformer` component.

## Actions
  - [x] Investigate if `spacy pretrain` also supports pre-training a `Transformer` component (open issue on their GH if needed). If Yes, determine if we can use this command instead of the aforementioned script (a) from `huggingface`.

**Comment by pafonta** [comment]
### TL;DR

> if `spacy pretrain` also supports pre-training a Transformer component

No.

From the documentation:

_The impact of `spacy pretrain` varies, but it will usually be worth trying if you’re `not using a transformer` model._

### What is `spacy pretrain` then?

This lets improve, thanks to multi-task learning, the performances of components relying on a `spacy.Tok2Vec` embedding layer. Indeed, this adds a language modeling objective to initialize as good as possible the embedding layer on which the components rely.

From the documentation:

_The `spacy pretrain` command will take a specific subnetwork within one of your components, and add additional layers to build a network for `a temporary task that forces the model to learn something about sentence structure and word cooccurrence statistics`._

### Why `spacy pretrain` is orthogonal to using a `transformer` component?

If one could simplify and use an analogy, `spacy pretrain` follows the same goal as pretraining a Transformer model with a masked language modeling task. However, `spacy pretrain` offers a lighter (e.g. no GPUs required) method at a cost of simpler contextual embeddings (e.g. only one embedding per token).

In a nutshell, the business of `spacy pretrain` and the business of `spacy-transformers.TransformerModel` are two different things. There might be some confusion coming from having two similar written forms (`tok2vec` vs `Tok2Vec`) for related but different concepts in the `spaCy` documentation. Indeed:
- `tok2vec`, all lower case, refers to:
- - the concept of embedding tokens into vectors (e.g. 'share a [...] `tok2vec` model between multiple components'),
- - the pipeline component doing so (e.g `[components.tok2vec]`),
- - or the component sublayer doing so (e.g. `[components.ner.model.tok2vec]`).
- while `Tok2Vec`, in camel case, refers to an implementation of the concept (e.g. `spacy.Tok2Vec.v1`).

### Reference

https://spacy.io/usage/embeddings-transformers#pretraining-details

**Comment by FrancescoCasalegno** [comment]
Thanks @pafonta! 

It seems that for our use case using a `Transformer` backbone gives better results than using a `Tok2Vec`, so can we conclude that it doesn't look like `spacy pretrain` will be useful for us?

If Yes, then for me this Issue can be closed and it is Done.

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno,

> can we conclude that it doesn't look like `spacy pretrain` will be useful for us?

Yes.

> If Yes, then for me this Issue can be closed and it is Done.

Closing, then.
